### PR TITLE
Create EntryFilesAnalyzer class to analyze a set of entry files

### LINKED ISSUE #237: Rework SourceFile analysis strategy
Right now JS-X-Ray is only capable to scan one sourcefile by one sourcefile. The [Scanner](https://github.com/NodeSecure/scanner) package is currently responsible of listing and iterating all JavaScript files from a given NPM tarball (for example).

What's the issue with that?

```
index.js
src/
  other.js
test/
  foobar.js
```

In the example above we will scan every files. But in reality there is a high probability that `test/foobar.js` will never be executed (and it will also be the biggest vector of false positives).

My idea is to add a new strategy that will take **entry files** as input. We will then only scan files imported from these entry points.

Eventually, we could combine the two ways of doing things to ensure greater security while reducing false positives overall.

### PR DESCRIPTION
**Pull request by jean-michelet**
Fixes #237

### CODE REVIEW

**Comment by jean-michelet** [comment]
Oops, I forgot to add the types, I'll do this.

File: test/EntryFilesAnalyser.spec.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    const generator = entryFilesAnalyser.analyse([entryUrl]);
+
+    await assertReport(generator, entryUrl, true);
+    await assertReport(generator, new URL("deps/default.js", FIXTURE_URL), true);
+    await assertReport(generator, new URL("deps/default.cjs", FIXTURE_URL), true);
+    await assertReport(generator, new URL("deps/default.mjs", FIXTURE_URL), true);
+    await assertReport(generator, new URL("deps/default.node", FIXTURE_URL), true);
+    await assertReport(generator, new URL("deps/default.jsx", FIXTURE_URL), true);
+
+    await assertAllReportsYielded(generator);
```
Comment (fraxken) [inline test/EntryFilesAnalyser.spec.js:70]: Probably better do to something like that

```
const result = await Array.fromAsync(generator);

// assert result
```

Comment (fraxken) [reply test/EntryFilesAnalyser.spec.js:70]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/fromAsync

Comment (jean-michelet) [reply test/EntryFilesAnalyser.spec.js:70]: Because not part of Node LTS yet, what do you suggest?

Comment (fraxken) [reply test/EntryFilesAnalyser.spec.js:70]: For now that ok

**Comment by jean-michelet** [comment]
Any suggestion @PierreDemailly?

**Review by fraxken — APPROVED** [review]

File: test/fixtures/entryFiles/deps/default.mjs [hunk 2]
```diff
@@ -0,0 +1 @@
+import internalDep from 'internalDep';
```
Comment (PierreDemailly) [inline test/fixtures/entryFiles/deps/default.mjs]: As internal imports must be relative (with extension) in ESM, i'm not sure to understand here.
Should it be `import externalDep from 'externalDep';` ?
Should it be `import internalDep from './internalDep.js';` + need to create `internalDep.js` file ?
If i'm not wrong, there are no ESM-import tests, should we add some ?

Same question in `default.js` -> `require('internalDep')` what is intended ?

**Comment by PierreDemailly** [comment]
Should we handle dynamic imports ? (In this PR or in a future one)

**Review by PierreDemailly — APPROVED** [review]

**Comment by jean-michelet** [comment]
> Should we handle dynamic imports ? (In this PR or in a future one)

#259

**Review by fraxken — APPROVED** [review]

**Pull request merged**